### PR TITLE
OLP-902: Vote a proposal & integration test for create/fund/vote proposals

### PR DESCRIPTION
**Pull request by ws4charlie**
https://oneledger.atlassian.net/browse/OLP-902

### CODE REVIEW

File: action/governance/createProposal.go [hunk 1]
```diff
@@ -113,36 +122,52 @@ func runTx(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
...
+	if coinGoal.LessThanEqualCoin(coin) {
+		result := action.Response{
+			Events: action.GetEvent(createProposal.Tags(), "create_proposal_too_much_funds"),
+		}
+		return false, result
+	}
+
+	//Calculate Deadlines
+	fundingDeadline := ctx.Header.Height + options.FundingDeadline
+	votingDeadline := fundingDeadline + options.VotingDeadline
```
Comment (ws4charlie) [inline action/governance/createProposal.go:147]: set voting deadline in funding Tx

File: action/governance/fundProposal.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+		return false, result
+	}
+
+	//4. Check if the Proposal has reached funding goal, when this contribution is added
+	//   Change the state of the proposal to VOTING, if funding goal is met
+	fundingAmount := balance.NewAmountFromBigInt(fundProposal.FundValue.Value.BigInt())
+	currentFundsforProposal := governance.GetCurrentFunds(proposal.ProposalID, ctx.ProposalMasterStore.ProposalFund)
+	newAmount := fundingAmount.Plus(currentFundsforProposal)
+	if newAmount.BigInt().Cmp(proposal.FundingGoal.BigInt()) >= 0 {
+		proposal.Status = governance.ProposalStatusVoting
```
Comment (ws4charlie) [inline action/governance/fundProposal.go:160]: voting deadline goes here

File: action/governance/voteProposal.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	}
+
+	// Check if proposal is in VOTING status
+	if proposal.Status != gov.ProposalStatusVoting {
+		return false, action.Response{
+			Log: fmt.Sprintf("vote proposal failed, id= %v, proposal not in VOTING status", vote.ProposalID)}
+	}
+
+	// Check if proposal voting height is passed
+	if proposal.VotingDeadline < ctx.Header.Height {
```
Comment (ws4charlie) [inline action/governance/voteProposal.go:101]: <=

File: action/governance/voteProposal.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+	// Get validator's voting power
+	validator, err := ctx.Validators.Get(vote.ValidatorAddress)
+	if err != nil {
+		return false, action.Response{
+			Log: fmt.Sprintf("vote proposal failed, id= %v, validator not found", vote.ProposalID)}
+	}
+
+	// Add this vote to proposal vote store
+	pv := gov.NewProposalVote(vote.ValidatorAddress, vote.Opinion, validator.Power)
```
Comment (ws4charlie) [inline action/governance/voteProposal.go:114]: check with pavan in case of power change

File: app/context.go [hunk 5]
```diff
@@ -112,6 +112,7 @@ func newContext(logWriter io.Writer, cfg config.Server, nodeCtx *node.Context) (
 
 	ctx.jobStore = jobs.NewJobStore(cfg, ctx.dbDir())
 	ctx.lockScriptStore = bitcoin.NewLockScriptStore(cfg, ctx.dbDir())
+	ctx.proposalMaster = NewProposalMasterStore(ctx.chainstate)
```
Comment (ws4charlie) [inline app/context.go:115]: remove duplicate store

File: service/broadcast/service.go [hunk 6]
```diff
@@ -143,3 +180,24 @@ func (svc *Service) TxCommit(req client.BroadcastRequest, reply *client.Broadcas
 	*reply = out
 	return nil
 }
+
+// TxCommit returns when the transaction has been committed to a block.
+func (svc *Service) TxCommitMtSig(req client.BroadcastMtSigRequest, reply *client.BroadcastReply) error {
```
Comment (ws4charlie) [inline service/broadcast/service.go:185]: add // TO DO here for TxSync

File: action/governance/finalizeProposal.go [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+		return errors.Wrap(err, "Failed in rewarding Proposer")
+	}
+	//Bounty Program
+	bountyAddress := action.Address(ctx.ProposalMasterStore.Proposal.GetOptions().BountyProgramAddr)
+	err = ctx.Balances.AddToAddress(bountyAddress, getPercentageCoin(c, totalFunding, proposalDistribution.BountyPool))
+	if err != nil {
+		return errors.Wrap(err, "Failed in adding to Adding to bounty program")
+	}
+	//Burn
+	burnAmount := getPercentageCoin(c, totalFunding, proposalDistribution.Burn)
```
Comment (pav-kap) [inline action/governance/finalizeProposal.go:201]: Are we doing something for burn here?

Comment (ws4charlie) [reply action/governance/finalizeProposal.go:201]: > Are we doing something for burn here?

I accidentally included this source file in my branch and I'll exclude it. Tanmay is still working on this Tx and I'll let him know your comments so he can take this into consideration.

File: action/governance/fundProposal.go [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+		for _, v := range validatorList {
+			vote := governance.NewProposalVote(v.Address, governance.OPIN_UNKNOWN, v.Power)
+			err = ctx.ProposalMasterStore.ProposalVote.Setup(proposal.ProposalID, vote)
+			if err != nil {
+				return false, action.Response{Log: "setup voting validator failed"}
+			}
+		}
+
+		//7. Update proposal status to VOTING
+		err = ctx.ProposalMasterStore.Proposal.Set(proposal)
```
Comment (KevinOneLedger) [inline action/governance/fundProposal.go:181]: Can you make sure you are setting the proposal to the active prefix?

Comment (kingpinXD) [reply action/governance/fundProposal.go:181]: This change is already there in my branch ,maybe I can put a comment to check it when we merge my branch to develop

**Pull request merged**